suse_migration_services/migration_config.py

```diff
@@ -31,6 +35,11 @@ class MigrationConfig(object):
     live image system. It is usually provided as a pre-setup
     version as part of the live migration image build.
     """
+    target_products = {
+        'ses': 'SLES/15.1/x86_64',
+        'sles': 'SLES/15/x86_64'
+    }
```

Comment:
I think we should not do that mapping here. The solution should be generic. This means in our image we pre-configure ```migration_product: _correct_value_```. The customer will have an opportunity to provide its own value by an overwrite and is allowed to shot himself. If we maintain a static list here the concept becomes inflexible. I suggest we take 1:1 what is written in the config file and do not map the value against a static mapping table

Reply:
I agree that the static list is less flexible and that customer should be able to set their own value for migration product. Following [this](https://github.com/SUSE/suse-migration-services/issues/73) conversation, one option is to check if the value provided matches any of the list else let the migration go with whatever value comes from the customer config (which would make the migration crash if the value is not valid).

Reply:
The point of the mapping is that we get to control where it goes. If you open it up as "free form" we open a trap. So we end up with "I set SLES/15.4/x86_64 and started at SLES 12" and we'll end up telling the customer "sorry not supported". Fair enough but certainly not satisfactory from the customer point of view. Or we end up having a humongous test matrix to make sure things work. I strongly recommend against opening the trap and with it the stream of support requests. Support is expensive for us and we can avoid it by locking things in place.

We can put the triplet into the config, that would help when we have to support migration for other architectures. But we'd still have to check the triplet for "is this a supported target". A "free form do what you want" approach is IMHO a giant mistake.

Reply:
If an upgrade to the specified target is possible or not is defined by the response of the zypper migration plugin. It will respond to you with either repos that can upgrade to the requested target or it will fail.

There is no good reason why we should maintain that mapping table to provide an error message "sorry can't do this" which is going to be sent by the zypper migration plugin asking for an upgrade path.

For the upcoming implementation of a simple "zypper dup" there is even no target spec to be set because the origin repo setup that a customer has to do beforehand will be the driver of an update.

So I don't think my suggestion is a mistake. Handling the request for yes we can upgrade this or not is the responsibility of the zypper migration plugin. You told me support matters. Yes it does and I rather tell the customer look this is what SCC told me: what you want is not possible rather than look this is our stupid mapping table that told you we can't do it but maybe we could.

Sorry I don't want a mapping table it's just another source of trouble in my eyes

Reply:
You're L3, we'll find out. Well, actually we probably will not find out as we usually don't hear back from support even if we put them into a tough position, but we'll probably hear from our partners.

Depending on SCC/SMT/RMT is a problem as there is no rule that enforces our support policies. For example, on a SLES 12 SP1 system I can run `zypper migration` and SLES 12 SP4 will be offered as a migration target. This is NOT supported. From SLES 12 SP1 the target that is supported, as in you will get help if things go wrong, is SLES 12 SP3.

Of course you will point out that this problem needs to be solve elsewhere, which is correct, and I agree. However, IMHO, as long as the problem exists letting the user walk into a trap when it can be prevented is not "the right thing to do".

I am not here to argue the point, I am just here to provide a different perspective and that maybe we should consider the potential user issues over "there is a workaround in the code to help the user avoid major issues". But if you think the work-around needs to go so we can be "technically correct" and let the user as you stated "shoot themselves in the foot" (paraphrased) then so be it.

Reply:
https://bugzilla.suse.com/show_bug.cgi?id=1135243

Reply:
> Handling the request for yes we can upgrade this or not is the responsibility of the zypper migration plugin

I agree that the responsibility whether a path is valid for migration or not belongs to zypper-migration plugin. My preference is 80% to let user provide whatever path and zypper-migration to decide and also to give the customer a nice experience but probably both of you have seen more similar cases and faced what can go wrong. 

So, as Robert did opening the bug, that invalid (available and not supported) paths should be fixed where appropriate. In the meantime, as I don't know how long it will take to have that fixed and I don't know what the consequences (test matrix, user experience, and so on Robert mentioned) of opening that door of let the customer migrate to whatever target they want are; we could compromise and make sure that the path is supported, maybe not with the mapping table but some kind of control until that bug is fixed, then update the code as removing that control and let the user provide any target they want to migrate and let zypper-migration plugin decides, as it is its responsibility.

One possible outcome of that approach is customers, who think a path is available because of a NOT supported path has been offered to them, end up thinking distro migration tool has limitations because it does not support that path. 

Thoughts?

Reply:
It is a trade off, as pretty much everything we do is.

On the one hand we have the creation of potentially letting the user choose something that is not supported, on the other hand we have the creation of the impression that the migration tool has limitations. But if we are honest with ourselves the migration tool does have limitations and we are OK with that, I think. We did not want to implement a solution that "solves all migration problems in the world". We set out to solve one use case.

Also we have the trade off between being "technically correct" and "creating a safe environment for the user".

All things considered we are probably in a 50/50 situation and no matter which path we choose there will be second guessing. My personal background would favor the route that reduces the risk for the user. But expecting the user to "know what hey are doing" is equally valid.

If the translation table is not the preferred direction and we want to let the user specify the triplet then at the very least I would put a look up table of supported values in place and compare the triplet. Using the triplet now will have the advantage that the format will not have to change when we need to support more than x86_64.

Yes, implementing the lookup table with verification we take some responsibility as to what we allow the user to do and what not. I don't think this is a bad thing in this case. But again it is a 50/50 decision.

Reply:
Not happy with the dictionary for targets but from user point of view, it makes sense to show them that it's not supported and avoid possible strange states left by a failed migration.

Reply:
From my point of view the mapping table just introduces product names that we have to own. It's a 1:1 mapping that does not more. The context is a translation from nameA to nameB and the benefit that we can react on an unknown name is something I don't see as a value. If a product name is set that is not known by the upgrade path in SCC the request response includes a well formulated error message saying that this product doesn't exist. From my perspective this is good enough and not considered a "strange state". The migration will fail without having any change or damage to the system. In default mode we will reboot and the user has a log file including an understandable error message.

I really don't see the need for the mapping table and it's not the job of the migration services to check on the integrity of the data provided for the migration. That's the job of the migration plugin. If you want to maintain a mapping table then do it there but it really makes no sense if the request answer does the validation on the server.

my 2cents. Less code to maintain is better